### PR TITLE
Update sync to return non-zero exit code on dry run

### LINKED ISSUE #1166: Feature: pip-sync --dry-run should return a non-zero exit code if changes were to occur
#### What's the problem this feature will solve?
I'm looking to add a pre-commit hook to check the environment is up to date

```
-   repo: local
    hooks:
    -   id: pip-sync
        name: pip-sync check
        entry: pip-sync --dry-run
        language: system
        always_run: true
        pass_filenames: false
```

#### Describe the solution you'd like
```
$ pip-sync --dry-run
Would install:
  numpy==1.18.5
$ $?
2
```

#### Alternative Solutions
various | awk stuff

#### Additional context
NA

### PR DESCRIPTION
**Pull request by francisbrito**
Resolves #1166

**Description**
Update sync function and CLI script so that they return non-zero exit codes if they're `dry-run` option is given.

**Changelog-friendly one-liner**: Update sync to return non-zero exit code on dry run

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by francisbrito** [comment]
Notes & suggestions:

* In order to improve legibility and avoid making mistakes, I'd suggest that exit codes be abstracted into module-level variables. This way, tests and script code can be refactored to use said variables instead of relying on "magic numbers", e.g:

```python
EXIT_CODE_SUCCESS = 0
EXIT_CODE_ERROR = 2
``` 

* The exit code to be returned when `--dry-run` is provided is `2`. This is in line with what was requested in #1166. If it makes more sense to return an alternative exit code for this case in particular. Please let me know so.

File: tests/test_cli_sync.py [hunk 1]
```diff
@@ -231,3 +231,16 @@ def test_sync_ask_accepted(check_call, runner):
     runner.invoke(cli, ["--ask", "--dry-run"], input="y\n")
 
     assert check_call.call_count == 2
+
+
+@mock.patch("piptools.sync.check_call")
+def test_sync_dry_run_returns_non_zero_exit_code(check_call, runner):
```
Comment (atugushev) [inline tests/test_cli_sync.py:237]:
```suggestion
def test_sync_dry_run_returns_non_zero_exit_code(runner):
```

I wonder, do we actually need to patch `check_call`?

Comment (francisbrito) [reply tests/test_cli_sync.py:237]: Oh, no, actually it is not required as the `sync` won't call `check_call` if ran in dry mode. My bad.

File: piptools/sync.py [hunk 2]
```diff
@@ -181,8 +183,11 @@ def sync(
             for ireq in sorted(to_install, key=key_from_ireq):
                 click.echo("  {}".format(format_requirement(ireq)))
 
+        exit_code = 2
```
Comment (atugushev) [inline piptools/sync.py:186]:
```suggestion
        exit_code = 1
```

code `1` would be more appropriate to not clash with code `2`, which is used by click in the [usage error context](https://github.com/pallets/click/blob/123dd717439d8620d8d6be5574d2c9f007952326/click/exceptions.py#L51).

File: tests/test_cli_sync.py [hunk 3]
```diff
@@ -231,3 +231,15 @@ def test_sync_ask_accepted(check_call, runner):
...
+def test_sync_dry_run_returns_non_zero_exit_code(runner):
+    """
+    Make sure non-zero exit code is returned when --dry-run is given.
+    """
+    with open("requirements.txt", "w") as req_in:
+        req_in.write("small-fake-a==1.10.0")
+
+    out = runner.invoke(cli, ["--dry-run"])
+
+    assert out.exit_code == 2
```
Comment (atugushev) [inline tests/test_cli_sync.py:245]:
```suggestion
    assert out.exit_code == 1
```

code `1` would be more appropriate to not clash with code `2`, which is used by click in the [usage error context](https://github.com/pallets/click/blob/123dd717439d8620d8d6be5574d2c9f007952326/click/exceptions.py#L51).

**Comment by atugushev** [comment]
Hey @francisbrito!

Thanks for your contribution! I took the liberty to address review comments and push changes to your branch since I'm going to release it soon. Feel free to fix "magic numbers" in a follow-up PR.

File: tests/test_cli_sync.py [hunk 4]
```diff
@@ -120,7 +120,7 @@ def test_merge_error(req_lines, should_raise, runner):
         assert out.exit_code == 2
         assert "Incompatible requirements found" in out.stderr
     else:
-        assert out.exit_code == 0
+        assert out.exit_code == 2
```
Comment (atugushev) [inline tests/test_cli_sync.py:123]:
```suggestion
        assert out.exit_code == 1
```

**Comment by atugushev** [comment]
Thanks @francisbrito for your first contribution! Good job!

**Comment by francisbrito** [comment]
@atugushev You're welcome!

(PS: Sorry for taking so long to reply, had a busy past week)

**Comment by pmi123** [comment]
I know I am late to the game here. However, I don't see the reason for always returning the error code 1 when `--dry-run` is called. As a comparison, `rsync --dry-run` returns 0 if no errors are detected, and only returns an error code if there is an actual error (e.g. invalid path). 

IMO, the purpose of a `--dry-run` option is to perform a test to see if the actual command and parameters will work. As long as there are no errors thrown, the `--dry-run` should return 0, just as if one had called `pip-sync` or `rsync` without `--dry-run`. Calling `--dry-run` is not an error condition in and of itself. 

Always returning an error code just because `--dry-run` was called eliminates the ability to test whether the `pip-sync` command will work as expected in a script. How does a script know if the error is from the actual command and parameters (e.g. invalid requirements.txt file) versus the error is just the default return value?

Mark

**Pull request merged**